// radiohead-serial/dist/src/RHDatagram.js
"use strict";
/*
 * Node.js module radiohead-serial
 *
 * RadioHead Library (http://www.airspayce.com/mikem/arduino/RadioHead/)
 * Copyright (c) 2014 Mike McCauley
 *
 * Port from native C/C++ code to TypeScript
 * Copyright (c) 2017 Peter Müller <peter@crycode.de> (https://crycode.de/)
 */
/// <reference types="node" />
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var radiohead_serial_1 = require("./radiohead-serial");
/**
 * Manager class for addressed, unreliable messages.
 */
var RHDatagram = (function (_super) {
    __extends(RHDatagram, _super);
    /**
     * Constructor
     * @param  {RH_Serial} driver      The used driver.
     * @param  {number}    thisAddress The address of this node.
     */
    function RHDatagram(driver, thisAddress) {
        var _this = _super.call(this) || this;
        _this._driver = driver;
        _this._thisAddress = thisAddress;
        return _this;
    }
    /**
     * Initialise this instance and the driver connected to it.
     * @return {Promise} Promise which will be resolved if the init of the driver is done.
     */
    RHDatagram.prototype.init = function () {
        var _this = this;
        return this._driver.init()
            .then(function () {
            _this.setThisAddress(_this._thisAddress);
            // emit recv event from driver on this class
            _this._driver.on('recv', function (msg) {
                _this.emit('recv', msg);
            });
        });
    };
    /**
     * Sets the address of this node. Defaults to 0.
     * This will be used to set the FROM address of all messages sent by this node.
     * In a conventional multinode system, all nodes will have a unique address.
     * @param {number} thisAddress The address of this node.
     */
    RHDatagram.prototype.setThisAddress = function (thisAddress) {
        this._driver.setThisAddress(thisAddress);
        // Use this address in the transmitted FROM header
        this.setHeaderFrom(thisAddress);
        this._thisAddress = thisAddress;
    };
    /**
     * Sends a message to the node(s) with the given address.
     * RH_BROADCAST_ADDRESS is a valid address which will cause the message to be
     * accepted by all RHDatagram nodes within range.
     * @param  {Buffer}  data    The buffer containing the data to send.
     * @param  {number}  len     Number of bytes from the buffer to send.
     * @param  {number}  address The address to send the message to.
     * @return {Promise}         Promise which will be resolved if sending is completed.
     */
    RHDatagram.prototype.sendto = function (data, len, address) {
        this.setHeaderTo(address);
        return this._driver.send(data, len);
    };
    /**
     * Returns the address of this node.
     * @return {number} The address of this node.
     */
    RHDatagram.prototype.thisAddress = function () {
        return this._thisAddress;
    };
    /**
     * Sets the TO header to be sent in all subsequent messages.
     * @param {number} to The new TO header value.
     */
    RHDatagram.prototype.setHeaderTo = function (to) {
        this._driver.setHeaderTo(to);
    };
    /**
     * Sets the FROM header to be sent in all subsequent messages.
     * @param {number} from The new FROM header value.
     */
    RHDatagram.prototype.setHeaderFrom = function (from) {
        this._driver.setHeaderFrom(from);
    };
    /**
     * Sets the ID header to be sent in all subsequent messages.
     * @param  {number} id The new ID header value.
     */
    RHDatagram.prototype.setHeaderId = function (id) {
        this._driver.setHeaderId(id);
    };
    /**
     * Sets and clears bits in the FLAGS header to be sent in all subsequent messages.
     * @param  {number} set   Bitmask of bits to be set.
     * @param  {number} clear Bitmask of flags to clear.
     */
    RHDatagram.prototype.setHeaderFlags = function (set, clear) {
        if (clear === void 0) { clear = radiohead_serial_1.RH_FLAGS_APPLICATION_SPECIFIC; }
        this._driver.setHeaderFlags(set, clear);
    };
    return RHDatagram;
}(events_1.EventEmitter));
exports.RHDatagram = RHDatagram;
