// radiohead-serial/dist/test/radiohead-serial-test.js
"use strict";
/*
 * Node.js module radiohead-serial
 *
 * Copyright (c) 2017 Peter Müller <peter@crycode.de> (https://crycode.de/)
 *
 * Node.js module for communication between some RadioHead nodes and Node.js using
 * the RH_Serial driver and the RHReliableDatagram manager of the RadioHead library.
 *
 *
 * RadioHead Library (http://www.airspayce.com/mikem/arduino/RadioHead/)
 * Copyright (c) 2014 Mike McCauley
 *
 * Port from native C/C++ code to TypeScript
 * Copyright (c) 2017 Peter Müller <peter@crycode.de> (https://crycode.de/)
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var mocha_typescript_1 = require("mocha-typescript");
var expect = require("expect.js");
var child_process_1 = require("child_process");
var radiohead_serial_1 = require("../src/radiohead-serial");
var socat;
var tty1 = null;
var tty2 = null;
var rhs1 = null;
var rhs2 = null;
//////////////////////////////////////
// Prepare virtual tty's with socat //
//////////////////////////////////////
var SocatPrepare = (function () {
    function SocatPrepare() {
    }
    SocatPrepare.prototype['spawn socat'] = function (done) {
        socat = child_process_1.spawn('socat', ['-d', '-d', 'pty,raw,echo=0', 'pty,raw,echo=0']);
        socat.on('error', function (error) {
            console.log('Error spawning `socat -d -d pty,raw,echo=0 pty,raw,echo=0`!');
            console.log(error);
            done(error);
        });
        socat.stderr.on('data', function (data) {
            var parts = data.toString().split('\n');
            for (var i = 0; i < parts.length; i++) {
                var m = parts[i].match(/PTY is (\/dev.*)$/);
                if (m) {
                    if (tty1 === null) {
                        tty1 = m[1];
                    }
                    else if (tty2 === null) {
                        tty2 = m[1];
                        setTimeout(done, 50);
                    }
                }
            }
        });
    };
    SocatPrepare.prototype['check the tty\'s created by socat'] = function () {
        expect(tty1).to.match(/^\/dev\/[\/a-zA-Z]+\d+$/);
        expect(tty2).to.match(/^\/dev\/[\/a-zA-Z]+\d+$/);
    };
    return SocatPrepare;
}());
__decorate([
    mocha_typescript_1.test
], SocatPrepare.prototype, "spawn socat", null);
__decorate([
    mocha_typescript_1.test
], SocatPrepare.prototype, "check the tty's created by socat", null);
SocatPrepare = __decorate([
    mocha_typescript_1.suite('prepare virtual tty\'s with socat')
], SocatPrepare);
/////////////////////////////////////////
// Start two RadioHeadSerial instances //
/////////////////////////////////////////
var RHS_Start = (function () {
    function RHS_Start() {
    }
    RHS_Start.prototype['create rhs1 (address 0x01)'] = function (done) {
        rhs1 = new radiohead_serial_1.RadioHeadSerial(tty1, 9600, 0x01);
        rhs1.on('init-done', function () {
            done();
        });
    };
    RHS_Start.prototype['create rhs2 (address 0x02)'] = function (done) {
        rhs2 = new radiohead_serial_1.RadioHeadSerial(tty2, 9600, 0x02);
        rhs2.on('init-done', function () {
            done();
        });
    };
    return RHS_Start;
}());
__decorate([
    mocha_typescript_1.test
], RHS_Start.prototype, "create rhs1 (address 0x01)", null);
__decorate([
    mocha_typescript_1.test
], RHS_Start.prototype, "create rhs2 (address 0x02)", null);
RHS_Start = __decorate([
    mocha_typescript_1.suite('create the two RadioHeadSerial instances')
], RHS_Start);
///////////////////////////////
// Send and receive messages //
///////////////////////////////
var SendRecv = (function () {
    function SendRecv() {
    }
    SendRecv.prototype['from 0x01 to 0x02'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs2.removeListener('data', recvListener);
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs2.on('data', recvListener);
        // send data
        rhs1.send(0x02, sendData).catch(function (err) {
            done(err);
        });
    };
    SendRecv.prototype['from 0x02 to 0x01'] = function (done) {
        var sendData = new Buffer('Here we go!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs1.removeListener('data', recvListener);
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs1.on('data', recvListener);
        // send data
        rhs2.send(0x01, sendData).catch(function (err) {
            done(err);
        });
    };
    SendRecv.prototype['from 0x01 to 0x02 with limited data length'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs2.removeListener('data', recvListener);
            var sendDataPart = new Buffer('Hey');
            if (msg.data.compare(sendDataPart) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs2.on('data', recvListener);
        // send data
        rhs1.send(0x02, sendData, 3).catch(function (err) {
            done(err);
        });
    };
    SendRecv.prototype['from 0x01 to 0x02 with control characters'] = function (done) {
        var STX = 0x02;
        var ETX = 0x03;
        var DLE = 0x10;
        var SYN = 0x16;
        var sendData = new Buffer('abcdef'); // data to be sent
        sendData[1] = 0x02; // STX
        sendData[2] = 0x03; // ETX
        sendData[3] = 0x10; // DLE
        sendData[4] = 0x16; // SYN
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs2.removeListener('data', recvListener);
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs2.on('data', recvListener);
        // send data
        rhs1.send(0x02, sendData).catch(function (err) {
            done(err);
        });
    };
    SendRecv.prototype['from 0x01 to broadcast'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs2.removeListener('data', recvListener);
            if (msg.headerTo !== radiohead_serial_1.RH_BROADCAST_ADDRESS) {
                done(new Error('The headerTo does not RH_BROADCAST_ADDRESS'));
                return;
            }
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs2.on('data', recvListener);
        // send data
        rhs1.send(radiohead_serial_1.RH_BROADCAST_ADDRESS, sendData).catch(function (err) {
            done(err);
        });
    };
    SendRecv.prototype['from 0x01 to 0x42 should fail'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // send data
        rhs1.send(0x42, sendData).catch(function (err) {
            done();
        });
    };
    return SendRecv;
}());
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x01 to 0x02", null);
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x02 to 0x01", null);
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x01 to 0x02 with limited data length", null);
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x01 to 0x02 with control characters", null);
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x01 to broadcast", null);
__decorate([
    mocha_typescript_1.test
], SendRecv.prototype, "from 0x01 to 0x42 should fail", null);
SendRecv = __decorate([
    mocha_typescript_1.suite('send and receive messages')
], SendRecv);
var TestFunctions = (function () {
    function TestFunctions() {
    }
    TestFunctions.prototype['setAddress(0x05)'] = function () {
        rhs2.setAddress(0x05);
    };
    TestFunctions.prototype['thisAddress()'] = function (done) {
        if (rhs2.thisAddress() === 0x05) {
            done();
        }
        else {
            done(new Error('thisAddress() did not report the set address'));
        }
    };
    TestFunctions.prototype['setRetries(7)'] = function () {
        rhs1.setRetries(7);
    };
    TestFunctions.prototype['getRetries()'] = function (done) {
        if (rhs1.getRetries() === 7) {
            done();
        }
        else {
            done(new Error('getRetries() did not report the set number of retries'));
        }
    };
    TestFunctions.prototype['setTimeout(100)'] = function () {
        rhs1.setTimeout(100);
    };
    TestFunctions.prototype['getRetransmissions()'] = function () {
        var retransmissions = rhs1.getRetransmissions();
        expect(retransmissions).to.be.a('number');
    };
    TestFunctions.prototype['resetRetransmissions()'] = function () {
        rhs1.resetRetransmissions();
    };
    TestFunctions.prototype['setPromiscuous(true)'] = function () {
        rhs1.setPromiscuous(true);
    };
    return TestFunctions;
}());
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "setAddress(0x05)", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "thisAddress()", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "setRetries(7)", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "getRetries()", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "setTimeout(100)", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "getRetransmissions()", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "resetRetransmissions()", null);
__decorate([
    mocha_typescript_1.test
], TestFunctions.prototype, "setPromiscuous(true)", null);
TestFunctions = __decorate([
    mocha_typescript_1.suite('test functions')
], TestFunctions);
///////////////////////////////
// Send and receive messages //
///////////////////////////////
var SendRecv2 = (function () {
    function SendRecv2() {
    }
    SendRecv2.prototype['from 0x01 to 0x05'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs2.removeListener('data', recvListener);
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs2.on('data', recvListener);
        // send data
        rhs1.send(0x05, sendData).catch(function (err) {
            done(err);
        });
    };
    SendRecv2.prototype['from 0x05 to 0x42 (0x01 with promiscuous mode)'] = function (done) {
        var sendData = new Buffer('Hey beauty!'); // data to be sent
        // function for received messages on rhs2
        var recvListener = function (msg) {
            // remove the listener function from the emitter
            rhs1.removeListener('data', recvListener);
            if (msg.data.compare(sendData) === 0) {
                done();
            }
            else {
                done(new Error('The received data does not match the sent data'));
            }
        };
        // attach the listener function to the emitter
        rhs1.on('data', recvListener);
        // send data
        rhs2.send(0x42, sendData).catch(function (err) {
            done(err);
        });
    };
    return SendRecv2;
}());
__decorate([
    mocha_typescript_1.test
], SendRecv2.prototype, "from 0x01 to 0x05", null);
__decorate([
    mocha_typescript_1.test
], SendRecv2.prototype, "from 0x05 to 0x42 (0x01 with promiscuous mode)", null);
SendRecv2 = __decorate([
    mocha_typescript_1.suite('send and receive messages')
], SendRecv2);
/////////////
// Cleanup //
/////////////
var Cleanup = (function () {
    function Cleanup() {
    }
    Cleanup.prototype['kill socat'] = function (done) {
        socat.on('close', function (code) {
            done();
        });
        socat.kill();
    };
    return Cleanup;
}());
__decorate([
    mocha_typescript_1.test
], Cleanup.prototype, "kill socat", null);
Cleanup = __decorate([
    mocha_typescript_1.suite('cleanup')
], Cleanup);
